x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/NativeAnalyticsProcess.java

```diff
@@ -85,12 +85,16 @@ public void restoreState(Client client, String stateDocIdPrefix) throws IOExcept
                     .setSize(1)
                     .setQuery(QueryBuilders.idsQuery().addIds(stateDocIdPrefix + ++docNum))
                     .get();
-                if (stateResponse.getHits().getHits().length == 0) {
-                    break;
+                try {
+                    if (stateResponse.getHits().getHits().length == 0) {
+                        break;
+                    }
+                    SearchHit stateDoc = stateResponse.getHits().getAt(0);
+                    logger.debug(() -> format("[%s] Restoring state document [%s]", config.jobId(), stateDoc.getId()));
+                    StateToProcessWriterHelper.writeStateToStream(stateDoc.getSourceRef(), restoreStream);
```

Comment:
@droberts195 could you confirm here as well? Writing to the `restoreStream` here doesn't require any references to the response once finished?